localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp

```diff
@@ -308,8 +319,9 @@ void NDTScanMatcher::callback_regularization_pose(
   regularization_pose_msg_ptr_array_.push_back(pose_conv_msg_ptr);
 }
 
-void NDTScanMatcher::callback_sensor_points(
-  sensor_msgs::msg::PointCloud2::ConstSharedPtr sensor_points_sensorTF_msg_ptr)
+void NDTScanMatcher::callback_points(
```

Comment:
How about making another module like `map_module` or `pose_initialization_module`? e.g. a module that only have the functionality of
- subscribe de-grounded scans
- calculate the proposed score and publish

However, separating the subscription may be tricky in terms of synchronization (since it may not make sense to use de-grounded scans whose timestamps are not close enough to the original scans that is used in NDT scan matching). I would also like to here opinions from others, if you have any. @KYabuuchi @YamatoAndo

Reply:
What about implementing a ground filter in the ndt_scan_matcher node?
For example, how about considering LiDAR point clouds with z-coordinate values below a certain value as ground point clouds?
I don't think an exact ground filter is necessary, so I think it is no problem to use an approximated ground filter.
This would also reduce the computational cost and solve the synchronization problem since there is no need to subscribe to a de-grounded scan.

(As an aside, it would also be interesting to look at the ND voxels generated from the map point cloud and extract the ground area to classficate the lidar point cloud.  Another good idea would be to use RANSAC.)

Reply:
I already make rough modifications to the code in order to ensure that both mechanisms work as follows.

1. Localization and calculate scores only using original point cloud.
2. Use the original point cloud for localization and use the de-grounded point cloud to calculate the score

The way it is to use a flag to determine whether to generate and use a de-grounded point cloud. If the flag is false, NDT matches using the original `callback_sensor_points` function, and if the flag is true, the `callback_points function` (I defined) is used and **both point clouds must be received simultaneously**.

The disadvantage of this modification is that two similar callback functions are defined and there is a lot of code duplication.

Reply:
I think my modification method is bad.

Reply:
@kminoda  Thanks your advice, but synchronization are very important.

Reply:
@YamatoAndo Thanks, good idea, I forgot this method.
The first idea was to use the ground-removed point cloud for localization and score calculation, so to get a good de-grounded point cloud, the **raw point cloud** is processed in `localization.xml`.
Now just using the de-grounded point cloud to calculate the score and the ground should be removed directly in the NDT node(for example use the RANSAC method) .


However, there is a potential risk that using **pre-processed point cloud** for de-grounding may not yield satisfactory results, and may not reflect the advantage of de-grounded point clouds in the NDT matching calculation score, so the experiment must be redone to compare the score difference between the two point clouds.

Reply:
Thank you! I also think @.YamatoAndo's idea makes sense. Maybe we can use even more simple method than RANSAC (for example, removing all the points that satisfies `point.z < base_link.z + vehicle_height` ). 

It is true that the validation must be done since using pre-process pcd for degrounding may increase the risk as you've mentioned. But IMO we can merge the proposed score with some quick evaluation so that the Autoware users can optionally use it and check if the score is useful enough.